marlowe-dashboard-client/src/MainFrame/State.purs

```diff
@@ -45,81 +48,138 @@ mkMainFrame =
+  }
+
+mkInsideState :: PubKeyHash -> InsideState
+mkInsideState pubKeyHash =
+  { wallet: pubKeyHash
+  , menuOpen: false
+  , screen: ContractsScreen Running
+  , card: Nothing
   , on: true
   }
 
+emptyWalletNicknameKey :: WalletNicknameKey
+emptyWalletNicknameKey = Tuple "" ""
+
+defaultWalletDetails :: WalletDetails
+defaultWalletDetails = WalletDetails { userHasPickedUp: false }
+
 handleQuery :: forall a m. Query a -> HalogenM State Action ChildSlots Msg m (Maybe a)
 handleQuery (ReceiveWebSocketMessage msg next) = do
   case msg of
-    (WS.ReceiveMessage (Right (ClientMsg on))) -> assign _on on
+    (WS.ReceiveMessage (Right (ClientMsg on))) -> assign (_insideState <<< _on) on
     -- TODO: other matches such as update current slot or apply transaction
     _ -> pure unit
   pure $ Just next
 
 handleAction :: forall m. MonadAff m => Action -> HalogenM State Action ChildSlots Msg m Unit
 handleAction Init = do
-  mCachedContactsJson <- liftEffect $ getItem contactsLocalStorageKey
-  for_ mCachedContactsJson
+  mCachedWalletsJson <- liftEffect $ getItem walletsLocalStorageKey
+  for_ mCachedWalletsJson
     $ \json -> do
         let
-          contacts =
-            runExcept
-              $ do
-                  foreignJson <- parseJSON json
-                  decode foreignJson
-        case contacts of
-          Right cachedContacts -> assign (_contactState <<< _contacts) cachedContacts
+          wallets =
+            runExcept do
+              decodeJSON json
+        case wallets of
+          Right cachedWallets -> assign _wallets cachedWallets
           _ -> pure unit
+  mCachedWalletJson <- liftEffect $ getItem walletLocalStorageKey
+  for_ mCachedWalletJson
+    $ \json -> do
+        let
+          wallet =
+            runExcept do
+              decodeJSON json
+        case wallet of
+          Right cachedWallet -> assign _subState $ Right $ mkInsideState cachedWallet
+          _ -> pure unit
+
+-- outside actions
+handleAction (SetOutsideCard outsideCard) = assign (_outsideState <<< _card) outsideCard
 
-handleAction (ToggleOverlay overlay) = do
-  mCurrentOverlay <- use _overlay
-  case mCurrentOverlay of
-    Just currentOverlay
-      | overlay == currentOverlay -> assign _overlay Nothing
-    _ -> assign _overlay $ Just overlay
+-- TODO: generate wallet on the backend; for now just create a random number
+handleAction GenerateNewWallet = do
+  randomNumber <- liftEffect random
+  let
+    key = show $ randomNumber
+  modify_
+    $ (_newWalletNicknameKey <<< set _key) key
+    <<< (_outsideState <<< set _card) (Just PickupNewWalletCard)
+
+handleAction PickupNewWallet = do
+  newPubKeyHash <- use (_newWalletNicknameKey <<< _key)
+  handleAction AddNewWallet
+  handleAction $ PickupWallet newPubKeyHash
+
+handleAction (LookupWallet string) = do
+  wallets <- use _wallets
+  for_ (findIndex (\key -> fst key == string) wallets) $ \key -> assign (_outsideState <<< _card) $ Just $ PickupWalletCard key
```

Comment:
`string` is a bad name, isn't it `nickname`? If not them my point is even more valid as I don't know!

Reply:
`$ \key` => `\key` I think

Reply:
`string` may be the right name, because we need to check it against pubkeys as well as nicknames. I've added the pubkey lookup too, which this reminds me that I forgot about.